[NETBEANS-5799] : Add support for Pattern Matching for switch (Preview)

singh-akhilesh opened the pull request:
https://issues.apache.org/jira/browse/NETBEANS-5799

jlahoda commented on java/java.completion/test/unit/src/org/netbeans/modules/java/completion/JavaCompletionTask117FeaturesTest.java:
```diff
@@ -0,0 +1,115 @@
+    private String getLatestSource(){
+        return SourceVersion.latest().name().substring(SourceVersion.latest().name().indexOf("_")+1);
+    }
+    private static final List<String> EXTRA_OPTIONS = new ArrayList<>();
```
Please move the EXTRA_OPTIONS into the TestCompilerOptionsQueryImplementation, as was done in: https://github.com/apache/netbeans/commit/8ec77c94d70e94af015ecccc2327626954c82237

This is so that other tests are not so much effected by the completion tests.

singh-akhilesh replied:
Addressed review comment, moved EXTRA_OPTIONS to class TestCompilerOptionsQueryImplementation.

jlahoda commented on java/java.source.base/src/org/netbeans/api/java/source/TreeMaker.java:
```diff
@@ -267,6 +267,18 @@ public CaseTree Case(List<? extends ExpressionTree> patterns, Tree body) {
+     * @see com.sun.source.tree.CaseTree
+     * @since 2.39
+     */
+    public CaseTree CaseMultipleSwitchPatterns(List<? extends Tree> patterns, Tree body) {
```
Maybe just `CaseMultiplePatterns`? Just a suggestion.

Also - there should probably be the similar method taking statements instead of body.

jlahoda commented on java/java.source.base/src/org/netbeans/modules/java/source/save/Reformatter.java:
```diff
@@ -2809,6 +2815,41 @@ private Boolean scanBindingPattern(Tree node, Void p) {
+            try {
+                Tree gpt = TreeShims.getGuardedPattern(node);
+                if (gpt != null) {
+                    Name name = TreeShims.getBinding(gpt);
```
I am not sure if this works - the nested pattern in the guarded pattern can be parenthesized pattern. Probably should just call `scan` on `gpt`? (Which can't be null?)

singh-akhilesh replied:
Addressed code review comment. Added code to handle  ---- case (String s) && s.<any> -> stmt;

jlahoda commented on java/java.source.base/src/org/netbeans/modules/java/source/save/Reformatter.java:
```diff
@@ -2809,6 +2815,41 @@ private Boolean scanBindingPattern(Tree node, Void p) {
+                    Tree type = TreeShims.getBindingPatternType(gpt);
+                    scan(type, p);
+                    if (name != null) {
+                        removeWhiteSpace(IDENTIFIER);
```
It is not quite clear what `remoteWhitespace` does - would `space();` work?

singh-akhilesh replied:
space(); also worked here, but adding some extra black spaces after case labels. In some case wrapping extra blank line also doesn't worked. If you want, I'm fine to replace this with space();

jlahoda reviewed (commented):
Overall, not bad. I've left some comments. A few more general comments:
-there should be more tests - for reformatting, for the hint, for `VeryPretty`. E.g. the reformatting tests will help avoid regressions on future versions of JDK.
-I would suggest to (now or later) look at `CasualDiff` for cases with patterns/guarded patterns/parenthesized patterns, that is likely to show up somewhere.

lahodaj reviewed (approved):
Overall, looks good. One further comment: possibly instead of add `visitCase` to each relevant TreeScanner subclass, add it to `ErrorAwareTreeScanner`/`ErrorAwareTreePathScanner`.

singh-akhilesh commented:
> Overall, looks good. One further comment: possibly instead of add `visitCase` to each relevant TreeScanner subclass, add it to `ErrorAwareTreeScanner`/`ErrorAwareTreePathScanner`.

Addressed review comment- Removed visitCase method from individual classes and added to ErrorAwareTreeScanner/ErrorAwareTreePathScanner
